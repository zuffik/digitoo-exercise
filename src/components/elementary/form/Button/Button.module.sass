.btn
  position: relative

.inner
  transition: opacity 100ms

  &.innerHidden
    opacity: 0

.spinner
  display: block
  opacity: 0
  position: absolute
  left: 50%
  top: 50%
  transform: translate(-50%, -50%)
  transition: opacity 100ms

  &.spinnerVisible
    opacity: 1
