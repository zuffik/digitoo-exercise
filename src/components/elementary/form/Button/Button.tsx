import * as React from 'react';
import classNames from 'classnames';
import {Colors} from '../../../../services/types/lib/bootstrap/Colors';
import {Spinner} from '../../progress/Spinner/Spinner';
import styles from './Button.module.sass';

interface Props extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  color?: Colors;
  loading?: boolean;
  large?: boolean;
}

export const Button: React.FC<Props> = ({loading, color, large, ...props}: Props) => {
  const btnColor = color || 'primary';
  const disabled = loading || props.disabled;
  return (
    <button
      {...props}
      data-testid="btn"
      className={classNames(props.className, styles.btn, 'btn', `btn-${btnColor}`, large && 'btn-lg')}
      disabled={disabled}
    >
      <span
          data-testid="btn-label"
          className={classNames(styles.inner, {
          [styles.innerHidden]: loading,
        })}
      >
        {props.children}
      </span>
      <span
          data-testid="btn-spinner"
          className={classNames(styles.spinner, {
          [styles.spinnerVisible]: loading,
        })}
      >
        <Spinner size={20} />
      </span>
    </button>
  );
};
