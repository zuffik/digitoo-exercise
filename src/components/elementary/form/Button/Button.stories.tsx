import * as React from 'react';
import {Button} from './Button';
import {boolean, select} from '@storybook/addon-knobs';

export default {
    title: 'Elementary/Form',
};

export const button = () => (
    <Button
        color={select(
            'color',
            {
                primary: 'primary',
                secondary: 'secondary',
                success: 'success',
                danger: 'danger',
                warning: 'warning',
                info: 'info',
                light: 'light',
                dark: 'dark',
            },
            'primary'
        )}
        loading={boolean('loading', false)}
    >
        Button text
    </Button>
);
