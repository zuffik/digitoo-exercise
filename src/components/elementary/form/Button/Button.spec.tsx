import * as React from 'react';
import {render} from '@testing-library/react'
import {Button} from "./Button";

describe('<Button/>', () => {
    it('should render', () => {
        const {asFragment, getByTestId} = render(<Button color="primary"/>);
        expect(getByTestId('btn')).toHaveClass("btn");
        expect(getByTestId('btn-label').className).not.toMatch(/innerHidden/);
        expect(getByTestId('btn-spinner').className).not.toMatch(/spinnerVisible/);
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render with spinner', () => {
        const {asFragment, getByTestId} = render(<Button color="primary" loading/>);
        expect(getByTestId('btn')).toHaveClass("btn");
        expect(getByTestId('btn-label').className).toMatch(/innerHidden/);
        expect(getByTestId('btn-spinner').className).toMatch(/spinnerVisible/);
        expect(asFragment()).toMatchSnapshot();
    });
});
