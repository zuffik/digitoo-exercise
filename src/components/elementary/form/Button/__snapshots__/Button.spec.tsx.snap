// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should render 1`] = `
<DocumentFragment>
  <button
    class="btn btn btn-primary"
    data-testid="btn"
  >
    <span
      class="inner"
      data-testid="btn-label"
    />
    <span
      class="spinner"
      data-testid="btn-spinner"
    >
      <div
        class="spinner-border"
        size="20"
        style="width: 20px; height: 20px;"
      />
    </span>
  </button>
</DocumentFragment>
`;

exports[`<Button/> should render with spinner 1`] = `
<DocumentFragment>
  <button
    class="btn btn btn-primary"
    data-testid="btn"
    disabled=""
  >
    <span
      class="inner innerHidden"
      data-testid="btn-label"
    />
    <span
      class="spinner spinnerVisible"
      data-testid="btn-spinner"
    >
      <div
        class="spinner-border"
        size="20"
        style="width: 20px; height: 20px;"
      />
    </span>
  </button>
</DocumentFragment>
`;
